feat: display notification badge on user account image - WPB-11270

El-Fitz opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-11270" title="WPB-11270" target="_blank"><img alt="Story" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10815?size=medium" />WPB-11270</a>  [iOS] Start and follow team creation flow for personal users
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove the jira markers to link tickets automatically -->


### Issue

Display a notification badge on user account image when they aren't part of a team, API v7 is available, and they haven't yet seen their profile. Hide it once they've viewed their profile.


---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [x] Make sure you use the API for UI elements that support large fonts.
- [x] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [x] New UI elements have Accessibility strings for VoiceOver.

El-Fitz committed:
feat: display notification on user account image when they aren't part of a team, API v7 is available, and they haven't yet seen their profile

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift:
```diff
@@ -115,6 +124,25 @@ extension ConversationListViewController {
+                return nil
+            }
+
+            let userDefaults = PrivateUserDefaults<UserDefaultsKey>(userID: userSession.selfUser.remoteIdentifier)
+            let value = userDefaults.object(forKey: .didViewSelfProfile)
```
could be hidden under a convernience method:

var didViewSelfProfile: Bool {
get set...
}

El-Fitz replied:
Moved into a separate object

netbe reviewed (commented):
early review, left a couple comments

netbe commented on WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift:
```diff
@@ -0,0 +1,167 @@
+
+// MARK: -
+
+final class NotificationBadgeView: UIView {
```
but for now don't change it

El-Fitz replied:
No. I just reused and simplified the code from the availability view.
What took time is integrating it and properly updating its state.

caldrian commented on WireUI/Sources/WireDesign/Icons/Image+ImageResource.swift:
```diff
@@ -21,4 +21,5 @@ import SwiftUI
 public extension Image {
     static let close = Image(.close)
     static let info = Image(.info)
+    static let infoCircle = Image(.infoCircle)
```
I strongly encourage to not have feature specific image assets at a central point.
Please move it to the target where it is used.
If this should be a view component to be reused, please move it to `WireReusableUIComponents`.

netbe replied:
done

caldrian commented on WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift:
```diff
@@ -93,12 +93,16 @@ public final class AccountImageView: UIView {
         set { availabilityIndicatorView.backgroundViewColor = newValue }
     }
 
+    public var notifications: ProfileNotifications? {
```
Why not just a simple Bool like `isNotificationBadgeVisible`?

netbe replied:
replaced

netbe committed:
Update WireUI/Sources/WireDesign/Icons/UImage+ImageResource.swift

netbe committed:
Merge branch 'develop' into feat/individual-to-team-notification-badge

netbe committed:
code review

netbe committed:
code review

netbe committed:
remove storage on logout

netbe committed:
format

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift:
```diff
@@ -210,6 +241,19 @@ extension ConversationListViewController.ViewModel {
+        ) { [weak self] _ in
+            guard let self else { return }
+            let userDefaults = PrivateUserDefaults<UserDefaultsKey>(userID: userSession.selfUser.remoteIdentifier)
+            userDefaults.set(true, forKey: .didViewSelfProfile)
```
done

El-Fitz replied:
Moved into a separate object

netbe commented on WireUI/Sources/WireAccountImageUI/AccountImageView/ProfileNotifications.swift:
```diff
@@ -0,0 +1,21 @@
+
+public enum ProfileNotifications {
+    case many
+}
```
done

netbe committed:
format

El-Fitz commented on wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift:
```diff
@@ -115,6 +124,36 @@ extension ConversationListViewController {
+                    userDefaults.set(false, forKey: .didViewSelfProfile)
+                }
+
+                return (value as? Bool) == true
```
This doesn't work when the value has never been set

El-Fitz committed:
fix: show notification badge when '.didViewSelfProfile' is nil

El-Fitz committed:
fix: display notifications badge in sidebar account image view, centralise didViewSelfProfile state

El-Fitz committed:
fix: notification badge inner icon size, and clean up

El-Fitz commented:
Updated it for the sidebar.
Also extracted the flag's storage so we can be sure it's updated in the UserDefaults before the notification is sent to observers, and to avoid duplicating the logic and user defaults key.

https://github.com/user-attachments/assets/68d05f35-5ca1-4dba-a66a-26df780a1f7c


https://github.com/user-attachments/assets/2a082956-d69e-4ace-9dcb-c2a1a674f2f6

![Simulator Screenshot - iPhone 16 - 2024-12-12 at 08 48 12](https://github.com/user-attachments/assets/928acd0f-e836-4b7c-bdb7-8ccd9b5636b4)

El-Fitz committed:
style: fix formatting

caldrian commented on wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:
```diff
@@ -1082,6 +1082,7 @@ public final class SessionManager: NSObject, SessionManagerType {
         deleteTemporaryData()
 
         PrivateUserDefaults.removeAll(forUserID: account.userIdentifier, in: sharedUserDefaults)
+        PrivateUserDefaults.removeAll(forUserID: account.userIdentifier, in: .standard)
```
Why is this change needed?

netbe replied:
I remove the user default storage which includes the notification badge on logout.

caldrian reviewed (commented):
Please add at least one snapshot test where the notification dot can be seen.

caldrian reviewed (commented):
<img width="177" alt="Screenshot 2024-12-12 at 12 03 44" src="https://github.com/user-attachments/assets/62258655-9fc7-49fa-9e29-e90e26b9e3bb" />

There is some duplicate white border.
I think the inner one needs to go, it should be red as well.

caldrian commented on WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift:
```diff
@@ -232,9 +236,25 @@ public final class AccountImageView: UIView {
+        notificationBadgeView.translatesAutoresizingMaskIntoConstraints = false
+        addSubview(notificationBadgeView)
+        NSLayoutConstraint.activate([
+            notificationBadgeView.widthAnchor.constraint(
+                equalToConstant: 14
+            ),
+            notificationBadgeView.heightAnchor.constraint(equalToConstant: 14),
+            notificationBadgeView.centerYAnchor.constraint(equalTo: accountImageViewWrapper.topAnchor, constant: 3.5),
+            notificationBadgeView.centerXAnchor.constraint(
+                equalTo: accountImageViewWrapper.trailingAnchor,
+                constant: -3.5
+            )
```
Could you try to make the size of the dot relative, please?
I think the remaining code somehow scales properly. (The view is not always displayed in the same size, like between iPhone and iPad [sidebar])
![ezgif-4-a979cb42f5](https://github.com/user-attachments/assets/401eb97d-d334-477b-bd31-f41585ffb321)

caldrian replied:
Actually the border around the image doesn't seem to scale, but still, it would be nice.

caldrian commented on WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift:
```diff
@@ -232,9 +236,25 @@ public final class AccountImageView: UIView {
 
+        // view which renders the notification badge
+        notificationBadgeView.translatesAutoresizingMaskIntoConstraints = false
+        addSubview(notificationBadgeView)
```
Can't we just use a simple `UIImageView`?

caldrian commented on WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift:
```diff
@@ -0,0 +1,169 @@
+    private let backgroundView = UIView()
+
+    /// An image view for the icon
+    private let iconView = UIImageView(image: .info)
```
not urgent now, but in my opinion we should get rid of the extensions to `UIImage` and `Image`
```
public extension UIImage {
    static let infoCircle = UIImage(resource: .infoCircle)
}
```
because they almost don't add any value:
```suggestion
    private let iconView = UIImageView(image: .init(resource: .info))
```

caldrian replied:
Even worse for the SwiftUI variant:
```
let infoCircle = Image.infoCircle
```
vs
```
let infoCircle = Image(.infoCircle)
```

caldrian commented on wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift:
```diff
@@ -115,6 +120,18 @@ extension ConversationListViewController {
+                return false
+            }
+
+            return !SelfProfileViewsMonitor.shared.didViewSelfProfile
```
Please don't use singletons.

caldrian commented on wire-ios/Wire-iOS/Sources/UserInterface/MainController/ZClientViewController.swift:
```diff
@@ -772,10 +774,12 @@ final class ZClientViewController: UIViewController {
                 cachedAccountImage,
-                cachedAccountInfo.isE2EICertified
+                cachedAccountInfo.isE2EICertified,
+                showNotificationsBadge: shouldShowNotificationsBadge(user: user)
```
optional
```suggestion
                isNotificationsBadgeVisible: ...
```

caldrian commented on wire-ios/Wire-iOS/Sources/UserInterface/SelfProfile/SelfProfileViewController.swift:
```diff
@@ -162,6 +162,7 @@ final class SelfProfileViewController: UIViewController {
 
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
+        SelfProfileViewsMonitor.shared.onDidViewSelfProfile()
```
Please don't use singletons.
Ideally we want to test the `SelfProfileViewController` individually without side effects.
For that we usually declare protocols for any dependencies.
We also use Sourcery for auto-creating the mocks.

caldrian commented on wire-ios/Wire-iOS/Sources/UserInterface/MainController/SidebarViewControllerBuilder.swift:
```diff
@@ -34,10 +34,12 @@ struct SidebarViewControllerBuilder {
                     source: accountImage.mapToAccountImageSource(),
-                    availability: availability?.mapToAccountImageAvailability()
+                    availability: availability?.mapToAccountImageAvailability(),
+                    // TODO: [WPB-11968] Use the actual value based on team membership, available API version & UserDefaults,
```
Ticket WPB-11968 is marked as in code review. Is there another PR which solves this TODO?

El-Fitz replied:
The comment is outdated.

El-Fitz committed:
chore: remove leftover comment

netbe committed:
Merge branch 'develop' into feat/individual-to-team-notification-badge

El-Fitz committed:
fix: remove singleton, use image instead of multiple views

El-Fitz committed:
fix: replace notification badge view with a simple UIImageView

El-Fitz committed:
fix: hide & show notifications badge

netbe committed:
Merge branch 'develop' into feat/individual-to-team-notification-badge

netbe committed:
format

netbe committed:
fix tests

netbe committed:
fomat

netbe committed:
add tests for showNotifications

netbe committed:
format

netbe reviewed (approved):
Looks good to me,
@El-Fitz  I updated the tests to add the showNotifications snapshots

netbe committed:
fix compilation error

netbe committed:
fix crash on test

netbe committed:
fix reference as mock user as red accent color

caldrian reviewed (approved):
I think the dark mode icon is missing for the notification dot.
Maybe leave a TODO somewhere.
Good work!
